### DISCUSSION TITLE
Usability enhancements

### DISCUSSION THREAD

**Discussion by bverkron** [redimp/otterwiki#60] [Ideas]
So far I'm _loving_ otterwiki because it's simple (no complicated GUIs get in the way of my writing), it stores data in markdown (making it easy to port other tools or edit externally). As a replacement for Confluence I think it might do pretty well despite being extremely simple!

However, I've noticed a few usability issues that make certain activities cumbersome or can even cause loss of work while editing. I'm summarize in text here but can provide visual examples later if necessary.

1) Navigating to the attachments page while editing loses any unsaved work.
If I'm editing a page and paste in an image and then want to rename that image via the attachments page it loses any unsaved work and there is no warning. Luckily I discovered this in the demo so I haven't lost any real work but this could be a big problem during day to day use. As a starting point to avoid loss of work even just having a warning pop up before navigating away from the page and allowing the user to cancel the navigation would be helpful.

1) Renaming images requires manual updates to references in pages
Maybe it would be possible to have otter automatically update the image names in the editor (or already saved file) when an image is renamed?

1) No draft editing, must do a full commit to save work to avoid losing it
It would be nice if there was a way to 'save' a draft without doing a full commit. When working on longer and more complicated docs having to save and do a commit message frequently to save my work in progress is a bit tedious. Also since we're using the commit model for saving there are probably lots of times where I don't actually want to commit it and do a full save (and thus publish it) but I also don't want to lose my work in the editor.

1) Cannot undo after viewing the preview
If I've made edits and I want to preview them to decide if I'm going to keep them I lose the ability to undo wth CTRL/CMD+Z when I return to the editing view and have to manually revert everything. A possible solution could be to launch the preview mode in a new window. I can force that manually with shortcut keys or right-clicks but I don't always remember and it's clunky.

1) No live preview
I understand this is probably a lot more work and complexity to implement than a separate preview page but one thing I appreciated about some other wiki's I tried was the split pane live preview while editing. That would be stellar but honestly just some of the usability tweaks with the current model would be good too.

1) No page organization structure
Maybe I'm being too picky here but a way to organize the pages via the UI and having a list of them and their heiarchy on the left side would be great.

Overall I think otterwiki is by far one of the best self-hosted wikis for my needs out there and if some of the usability issues were fixed it would almost be my endgame for a wiki (as of right now anyway). Great work!

**Comment by redimp** [comment]
Thanks for the feedback, thats very helpful! I will address each individual point, eventually. Please be patient, I will start with 6.)

6.) Typically the `Home` serves as starting point and users build their page organization there, often with lists. I'm sure you have considered this already.

I was thinking about adding an area to the left sidebar where Admins can add links (to pages) and pick an icon (of Font Awesome). Maybe it would be useful if Users can individually customize their sidebar too.

**Reply by bverkron** [reply]
> I'm sure you have considered this already.

Actually I hadn't. I think mostly because most tools I've used for a long time have some kind of menu or list in them that is automatically updated as you create content. It feels like stepping back in time to have to manually edit and manage a list of anything. It would become very tedious to remember to create links in the home page for each new wiki page you create and then go do it. And if you renamed a page to them have to go update the reference. It's not a huge deal when the wiki is small but the larger it grows the more unwieldy and time consuming that becomes.

Mostly what I'm hoping for is a simple folder / organizational hierarchy on the left side similar to [Confluence](https://confluence.atlassian.com/doc/configure-the-sidebar-317194694.html), which is what I'd like to migrate away from. While having a powerful search should in theory cover most use cases there are times when, as a user, you are not sure of the right terms to use or can't put the concept into words and it's more effective to start with a broad topic and drill down in the hierarchy to find what you need.

<img width="309" alt="image" src="https://github.com/redimp/otterwiki/assets/19174503/12fd0161-2bb1-45d5-97ac-434e24f9cd6e">

What about having the ability to organize the markdown file into folders in the file system and have that act as the menu/list hierarchy on the left. Ideally you would have a way to do that in the UI too. You could have the list visibility be optional incase people want to only use the current home page / search method.

The A-Z index page does handle some of this but as it grows it's likely to become visually overwhelming and it's not easy to keep organized unless you do sort of categorized page titles to use as pseudo 'folders' and even then if you want to rename a 'folder' you would have to rename all the pages.

<img width="916" alt="image" src="https://github.com/redimp/otterwiki/assets/19174503/4cd8e308-514f-4c0a-a978-cb3b8405dbef">

> I was thinking about adding an area to the left sidebar where Admins can add links (to pages) and pick an icon (of Font Awesome). Maybe it would be useful if Users can individually customize their sidebar too.

That would also be really nice but I see that as separate from what I'm trying to describe. In hindsight I think my original description was a bit too vague.

**Reply by redimp** [reply]
Uhh, quick comment: You already can organize files in subfolders: e.g https://otterwiki.com/Examples

For your example try to rename the Pages to

- `Financial`
- `Financial/Insurance`
- `Financial/Weekly Budget`
- ..
- `Technology/Infrastructure/Docker Server`
- `Technology/Infrastructure/Docker Server/Wireguard`

But back to the sidebar: Always displaying the full wiki contents in the sidebar would it mak cluttered. From my observation on how people use confluence: The sidebar is so full, nobody uses it and relies on the search ... But what do I know?

I think it is a viable option, to let the Admin configure not only a sidebar, but also that the wiki structure is displayed (up to a confiurable depth).

**Comment by redimp** [comment]
### 1.) "Navigating to the attachments page while editing loses any unsaved work."

I agree, the wiki should warn you when you are navigating away and might lose what you've done. Yes you can return with the Back Button, but displaying a warning is the expected behaviour on web pages. Added issue #64. 

### 2.) "Renaming images requires manual updates to references in pages"

This would be a convient feature. Added as #65.

### 3.) "No draft editing, must do a full commit to save work to avoid losing it"

I was thinking about adding a feature like (auto) saving to drafts, so that when a user navigates back to a page with an existing draft this will be displayed and editing can be continued. The state of the draft will **not** be stored in the filesystem, though. This would need a proper commit.

Added as #66.

### 4.) "Cannot undo after viewing the preview"

How the switching between `Editor` and `Preview` currenlty works is far from perfect. I would like to change that in a way like it works in e.g. github. This would solve the issue with loosing the Undo/Redo history as well.

Added as #67.

### 5. "Live preview"

Live preview odr side by side editing is in most cases done where the markdown is rendered on the client side. We are rendering on the server side. Live preview would be very clunky. I don't think this will be fun. If you still want this, please open an issue for it.

**Comment by bverkron** [comment]
Sorry for the late reply, been sick the past week or two.

Appreciate you considering the feedback and opening issues to address them. #1 is my main concern since it can easily result in data loss. I think if that were resolved I could pretty comfortably implement Otter as my household wiki program and start porting over from Confluence.

The rest are more QoL improvements. #5 is totally optional if there were at least a better way to see the preview without losing the undo functionality because you are navigated away from the editor.

I have a ton of side projects on the go in addition to a family of 5 + dog _and_ chronic health problems... but if I find some spare time I can try to pitch in on development. Though I'm not too familiar with the languages, frameworks, etc that this project uses :(
